File: components/Analog_stage_output.py
import numpy as np
from circuit_solver.subcircuit import Subcircuit
from components.resistor import Resistor
from components.capacitor import Capacitor
from components.diode import Diode # Per la distorsione/limitazione


class AnalogOutputStage(Subcircuit):
    """
    Simula uno stadio di uscita analogico di un convertitore vintage.
    Include:
    - Filtro di ricostruzione passa-basso RC
    - Resistenza di uscita (Output Impedance)
    - Possibile limitazione/saturazione (via diodi)
    """
    def __init__(self, name: str, input_node: str, output_node: str, ground_node: str = '0',
                 filter_R: float = 1e3,          # Resistenza filtro di ricostruzione
                 filter_C: float = 100e-9,       # Capacità filtro di ricostruzione
                 output_resistance: float = 50.0, # Impedenza d'uscita (50 Ohm)
                 clipper_threshold: float = 0.7, # Soglia approssimativa di clipping per diodi
                 diode_saturation_current: float = 1e-9, # Is per diodo clipper
                 diode_emission_coefficient: float = 1.0): # N per diodo clipper
        
        super().__init__(name, input_nodes=[input_node], output_nodes=[output_node])

        # Nodi interni
        self.filter_in_node = f"{name}_filter_in"
        self.clipper_node = f"{name}_clipper_out"
        
        # Aggiungi i nodi al sottocircuito
        self.add_node(input_node)
        self.add_node(output_node)
        self.add_node(ground_node)
        self.add_node(self.filter_in_node)
        self.add_node(self.clipper_node)

        # 1. Filtro di Ricostruzione Passa-Basso (RC)
        # Questo serve a "lisciare" il segnale dopo il DAC digitale (non simulato qui)
        self.add_component(Resistor(f"{name}_FilterR_in", input_node, self.filter_in_node, filter_R))
        self.add_component(Capacitor(f"{name}_FilterC_out", self.filter_in_node, ground_node, filter_C))

        # 2. Stadio di Clipper/Saturazione (diode limiter)
        # Per simulare una leggera compressione o saturazione.
        # Nota: Questo è un clipper duro. Potresti voler usare qualcosa di più soft
        # come un OpAmp con feedback non lineare per un effetto più graduale.
        # Qui usiamo un diodo in parallelo per tagliare il segnale a un certo livello.
        # D2 è un diodo che clippa il segnale positivo.
        # D3 è un diodo che clippa il segnale negativo (collegato al contrario).
        # R_clipper_series limita la corrente attraverso i diodi.
        self.add_component(Resistor(f"{name}_R_clipper_series", self.filter_in_node, self.clipper_node, 100)) # Limita corrente
        
        # Diodi per clipping bidirezionale
        # Per controllare la soglia di clip:
        # Usa un valore alto di Is per simulare una tensione di soglia più bassa (vicino a 0.2-0.3V)
        # Oppure metti più diodi in serie per aumentare la soglia di ~0.7V per diodo.
        # Per clipper_threshold: useremo Rs per approssimare il livello di clipping desiderato.
        # Un diodo singolo si attiva intorno a 0.7V. Per clip a 0.3V, Is dev'essere più alta (es. 1e-6).
        # Se vogliamo una soglia specifica, dovremmo modellare con una sorgente di tensione in serie ai diodi.
        
        # Per simulare una soglia di clipping approssimativa con un diodo singolo (o parallelo)
        # usiamo Is per "spostare" la curva I-V. Valori più alti di Is significano che il diodo
        # conduce di più a tensioni più basse.
        
        # Diodo per clipping positivo (taglia i picchi sopra la soglia del diodo)
        self.add_component(Diode(f"{name}_D_pos", self.clipper_node, ground_node,
                                  Is=diode_saturation_current * (10**(clipper_threshold/0.026)) , # Ajusta Is per soglia desiderata (V_T=26mV)
                                  N=diode_emission_coefficient))
        # Diodo per clipping negativo (taglia i picchi sotto la soglia negativa del diodo)
        self.add_component(Diode(f"{name}_D_neg", ground_node, self.clipper_node,
                                  Is=diode_saturation_current * (10**(clipper_threshold/0.026)),
                                  N=diode_emission_coefficient))

        # 3. Resistenza di Uscita
        self.add_component(Resistor(f"{name}_Rout", self.clipper_node, output_node, output_resistance))

        print(f"Stadio di Uscita Analogico '{name}' creato. Filtro, impedenza d'uscita e clipping modellati.")

File: components/Analog_stage_output.cpp
#include "AnalogInputStage.h"
#include <iostream>
#include <algorithm> // Per std::find

// Implementazione del metodo addNode
void AnalogInputStage::addNode(const std::string& node_name) {
    // Aggiunge il nodo solo se non è già presente
    if (std::find(nodes_.begin(), nodes_.end(), node_name) == nodes_.end()) {
        nodes_.push_back(node_name);
    }
}

// Costruttore della classe AnalogInputStage
AnalogInputStage::AnalogInputStage(const std::string& name,
                                 const std::string& input_node,
                                 const std::string& output_node,
                                 const std::string& ground_node,
                                 double input_resistance,
                                 double gain_factor,
                                 double filter_R,
                                 double filter_C,
                                 double diode_saturation_current,
                                 double diode_emission_coefficient)
    : name_(name),
      input_nodes_({input_node}),
      output_nodes_({output_node}),
      ground_node_(ground_node)
{
    // Nodi interni specifici per questa istanza del sottocircuito
    gain_node_ = name_ + "_gain_out";
    dist_node_ = name_ + "_dist_out";
    filter_node_ = name_ + "_filter_out";
    diode_anode_node_ = name_ + "_d1_in";
    filter_mid_node_ = name_ + "_filter_mid";

    // Aggiungi i nodi al sottocircuito
    addNode(input_node);
    addNode(output_node);
    addNode(ground_node_);
    addNode(gain_node_);
    addNode(dist_node_);
    addNode(filter_node_);
    addNode(diode_anode_node_);
    addNode(filter_mid_node_);

    // 1. Resistenza di Ingresso
    // La resistenza di ingresso è tra input_node e il nodo di output dello stadio di guadagno
    addComponent<Resistor>(name_ + "_Rin", input_node, gain_node_, input_resistance);

    // 2. Stadio di Guadagno (semplice partitore resistivo per attenuazione/guadagno)
    if (gain_factor < 1.0) {
        // Calcola la resistenza in serie per ottenere l'attenuazione desiderata
        // Nota: questa è una semplificazione. In un partitore vero, avresti un'altra resistenza a massa
        // per creare il rapporto. Qui assumiamo che R_in e R_gain formino il partitore con il carico successivo.
        // Questa logica necessiterebbe di un aggiustamento per un partitore standard.
        // Ad esempio: Vout = Vin * (R_parallel / (R_series + R_parallel))
        // La tua logica python era R_gain_series = input_resistance / gain_factor - input_resistance.
        // Questo implica che R_in è la resistenza superiore e R_gain_series è la resistenza inferiore,
        // ma R_gain_series dovrebbe andare a massa per un partitore tradizionale.
        // Per mantenere la logica Python, lo interpretiamo come una resistenza in serie che modifica l'impedenza vista.
        
        // Se vogliamo un'attenuazione, l'idea è che la R_in faccia parte del partitore.
        // Per ottenere l'attenuazione 'gain_factor', consideriamo che l'impedenza di ingresso è data.
        // Se gain_factor < 1, significa che il segnale viene ridotto.
        // Un semplice modo per fare attenuazione passiva con resistori è un partitore.
        // R_upper = input_resistance (già definita)
        // R_lower = ?
        // Vout = Vin * (R_lower / (R_upper + R_lower)) = Vin * gain_factor
        // gain_factor = R_lower / (R_upper + R_lower)
        // R_lower = gain_factor * (R_upper + R_lower)
        // R_lower = gain_factor * R_upper + gain_factor * R_lower
        // R_lower * (1 - gain_factor) = gain_factor * R_upper
        // R_lower = (gain_factor * R_upper) / (1 - gain_factor)
        
        // Questo calcolo è per un partitore resistivo standard dove R_upper è tra Vin e Vout,
        // e R_lower è tra Vout e massa.
        // La tua logica Python era: r_gain_series = input_resistance / gain_factor - input_resistance.
        // Questo sembra voler modellare una resistenza in serie al segnale dopo R_in per arrivare a un certo punto.
        // Manteniamo la logica Python per coerenza.
        double r_gain_series = input_resistance / gain_factor - input_resistance;
        if (r_gain_series < 1e-6) r_gain_series = 1e-6; // Evita valori troppo piccoli o negativi
        addComponent<Resistor>(name_ + "_RGain", gain_node_, dist_node_, r_gain_series);
    } else {
        // Se guadagno >= 1, consideralo come passante diretto (resistenza molto piccola)
        addComponent<Resistor>(name_ + "_RPass", gain_node_, dist_node_, 1e-6); // Quasi un corto
    }

    // 3. Distorsione (diodo soft-clipping)
    // R_dist_series limita la corrente attraverso il diodo.
    addComponent<Resistor>(name_ + "_R_dist_series", dist_node_, diode_anode_node_, 1e3);
    // Il diodo è tra il suo anodo_node e il nodo di massa
    addComponent<Diode>(name_ + "_D1", diode_anode_node_, ground_node_, diode_saturation_current, diode_emission_coefficient);
    // R_dist_parallel è tra il nodo del diodo e il nodo di ingresso del filtro
    addComponent<Resistor>(name_ + "_R_dist_parallel", diode_anode_node_, filter_node_, 10e3);

    // 4. Filtro Anti-Aliasing Passa-Basso (RC)
    // La resistenza del filtro è tra l'uscita dello stadio di distorsione e il nodo intermedio del filtro
    addComponent<Resistor>(name_ + "_FilterR", filter_node_, filter_mid_node_, filter_R);
    // Il condensatore del filtro è tra il nodo intermedio e la massa
    addComponent<Capacitor>(name_ + "_FilterC", filter_mid_node_, ground_node_, filter_C);
    
    // Output del sottocircuito: collega il nodo di uscita del filtro all'output_node dichiarato
    addComponent<Resistor>(name_ + "_ROut", filter_mid_node_, output_node, 1e-6); // R quasi 0 per collegare a output_node

    std::cout << "Stadio di Ingresso Analogico '" << name_ << "' creato. Impedenze, guadagno e distorsione modellati." << std::endl;
}
